### PR TITLE
Provide effort to fix for complexity rules (S3776 and S1541)

### LINKED ISSUE #165: Provide effort to fix for complexity rules (S3776 and S1541)

### PR DESCRIPTION
**Pull request by vilchik-elena**
Fixes #165

### CODE REVIEW

File: sonarts-sq-plugin/sonar-typescript-plugin/src/main/java/org/sonar/plugin/typescript/ExternalTypescriptSensor.java [hunk 1]
```diff
@@ -292,38 +292,46 @@ private void saveIssues(SensorContext sensorContext, Issue[] issues, TypeScriptR
...
-            issue.endPosition.line + 1,
-            issue.endPosition.character));
-        }
+        saveIssue(sensorContext, typeScriptRules, issue, inputFile);
+      }
+    }
+  }
 
-        newIssue.at(location);
+  private void saveIssue(SensorContext sensorContext, TypeScriptRules typeScriptRules, Issue issue, InputFile inputFile) {
```
Comment (SonarTech) [inline sonarts-sq-plugin/sonar-typescript-plugin/src/main/java/org/sonar/plugin/typescript/ExternalTypescriptSensor.java]: Code Smell Code Smell: Make "saveIssue" a "static" method. ([squid:S2325](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2325))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=sonarts&branch=issue-165&issues=AWC7YjLnqDrCo5ysghE3&open=AWC7YjLnqDrCo5ysghE3)

[ISSUE_KEY]: AWC7YjLnqDrCo5ysghE3

File: sonarts-core/tests/runRule.ts [hunk 2]
```diff
@@ -119,13 +120,15 @@ function parseErrorsFromMarkup(source: string) {
       const startColumn = line.indexOf("^");
       const endColumn = line.lastIndexOf("^") + 1;
       const message = line.match(/\{\{(.+)\}\}/)[1];
-
+      const costMatch = line.match(/\[\[(\d+)\]\]/);
```
Comment (stas-vilchik) [inline sonarts-core/tests/runRule.ts]: I suggest to be more specific and use something like `[[cost:1]]`

File: sonarts-core/tests/runRule.ts [hunk 3]
```diff
@@ -197,9 +200,14 @@ function mapToLintErrors(failures: tslint.RuleFailure[]): LintError[] {
   return failures.map(failure => {
     const startPosition = failure.getStartPosition().getLineAndCharacter();
     const endPosition = failure.getEndPosition().getLineAndCharacter();
+    let cost: number;
```
Comment (stas-vilchik) [inline sonarts-core/tests/runRule.ts]: should be `number | undefined`

File: sonarts-core/tests/runRule.ts [hunk 4]
```diff
@@ -197,9 +200,14 @@ function mapToLintErrors(failures: tslint.RuleFailure[]): LintError[] {
...
+    if (isSonarIssue(failure)) {
+      cost = failure.getCost();
+    }
     return {
       endPos: { col: endPosition.character, line: lineNumberedFromOne(endPosition.line) },
       message: failure.getFailure(),
+      cost,
```
Comment (stas-vilchik) [inline sonarts-core/tests/runRule.ts]: I'd inline this like:
```typescript
cost: isSonarIssue(failure) ? failure.getCost() : undefined
```

File: sonarts-sq-plugin/sonar-typescript-plugin/src/main/java/org/sonar/plugin/typescript/ExternalTypescriptSensor.java [hunk 5]
```diff
@@ -394,6 +402,7 @@ private static File getChildDirectoryByName(File directory, String name) {
     String name;
     String ruleName;
     SecondaryLocation[] secondaryLocations;
+    Integer cost;
```
Comment (stas-vilchik) [inline sonarts-sq-plugin/sonar-typescript-plugin/src/main/java/org/sonar/plugin/typescript/ExternalTypescriptSensor.java]: why not `Double`? You cast it to `double` anyway: 
```java
newIssue.gap((double)issue.cost);
```

File: sonarts-core/tests/runRule.ts [hunk 6]
```diff
@@ -200,10 +200,7 @@ function mapToLintErrors(failures: tslint.RuleFailure[]): LintError[] {
   return failures.map(failure => {
     const startPosition = failure.getStartPosition().getLineAndCharacter();
     const endPosition = failure.getEndPosition().getLineAndCharacter();
-    let cost: number;
-    if (isSonarIssue(failure)) {
-      cost = failure.getCost();
-    }
+    let cost = isSonarIssue(failure) ? failure.getCost() : undefined;
```
Comment (stas-vilchik) [inline sonarts-core/tests/runRule.ts]: use `const`

**Pull request merged**